Parametric/Syntax/MTerm.agda

```diff
@@ -0,0 +1,192 @@
+      -- languages.
+      vThunk : ∀ {τ} → Comp Γ τ → Val Γ (U τ)
+      vConst : ∀ {Σ τ} →
+        (c : ValConst Σ τ) →
+        (args : Vals Γ Σ) →
+        Val Γ τ
+
+    data Comp Γ : (τ : CompType) → Set where
+      cConst : ∀ {Σ τ} →
+        (c : CompConst Σ τ) →
+        (args : Vals Γ Σ) →
+        Comp Γ τ
+
+      cForce : ∀ {τ} → Val Γ (U τ) → Comp Γ τ
+      cReturn : ∀ {τ} (v : Val Γ τ) → Comp Γ (F τ)
+      {-
+      -- Originally, M to x in N. But here we have no names!
+      _into_ : ∀ {σ τ} →
+        (e₁ : Comp Γ (F σ)) →
+        (e₂ : Comp (σ •• Γ) τ) →
+        Comp Γ τ
+      -}
+      -- The following constructor is the main difference between CBPV and this
+      -- monadic calculus. This is better for the caching transformation.
+      _into_ : ∀ {σ τ} →
+        (e₁ : Comp Γ (F σ)) →
+        (e₂ : Comp (σ •• Γ) (F τ)) →
+        Comp Γ (F τ)
+      cAbs : ∀ {σ τ} →
+        (t : Comp (σ •• Γ) τ) →
+        Comp Γ (σ ⇛ τ)
+      cApp : ∀ {σ τ} →
+        (s : Comp Γ (σ ⇛ τ)) →
+        (t : Val Γ σ) →
+        Comp Γ τ
+
+  weaken-val : ∀ {Γ₁ Γ₂ τ} →
+    (Γ₁≼Γ₂ : Γ₁ ≼≼ Γ₂) →
+    Val Γ₁ τ →
+    Val Γ₂ τ
+
+  weaken-comp : ∀ {Γ₁ Γ₂ τ} →
+    (Γ₁≼Γ₂ : Γ₁ ≼≼ Γ₂) →
+    Comp Γ₁ τ →
+    Comp Γ₂ τ
+
+  weaken-vals : ∀ {Γ₁ Γ₂ Σ} →
+    (Γ₁≼Γ₂ : Γ₁ ≼≼ Γ₂) →
+    Vals Γ₁ Σ →
+    Vals Γ₂ Σ
+
+  weaken-val Γ₁≼Γ₂ (vVar x) = vVar (weaken-val-var Γ₁≼Γ₂ x)
+  weaken-val Γ₁≼Γ₂ (vThunk x) = vThunk (weaken-comp Γ₁≼Γ₂ x)
+  weaken-val Γ₁≼Γ₂ (vConst c args) = vConst c (weaken-vals Γ₁≼Γ₂ args)
+  weaken-comp Γ₁≼Γ₂ (cConst c args) = cConst c (weaken-vals Γ₁≼Γ₂ args)
+  weaken-comp Γ₁≼Γ₂ (cForce x) = cForce (weaken-val Γ₁≼Γ₂ x)
+  weaken-comp Γ₁≼Γ₂ (cReturn v) = cReturn (weaken-val Γ₁≼Γ₂ v)
+  weaken-comp Γ₁≼Γ₂ (_into_ {σ} c c₁) = (weaken-comp Γ₁≼Γ₂ c) into (weaken-comp (keep σ •• Γ₁≼Γ₂) c₁)
+  weaken-comp Γ₁≼Γ₂ (cAbs {σ} c) = cAbs (weaken-comp (keep σ •• Γ₁≼Γ₂) c)
+  weaken-comp Γ₁≼Γ₂ (cApp s t) = cApp (weaken-comp Γ₁≼Γ₂ s) (weaken-val Γ₁≼Γ₂ t)
+
+  weaken-vals Γ₁≼Γ₂ ∅ = ∅
+  weaken-vals Γ₁≼Γ₂ (px • ts) = (weaken-val Γ₁≼Γ₂ px) • (weaken-vals Γ₁≼Γ₂ ts)
+
+  fromCBN : ∀ {Γ τ} (t : Term Γ τ) → Comp (fromCBNCtx Γ) (cbnToCompType τ)
+
+  fromCBNTerms : ∀ {Γ Σ} → Terms Γ Σ → Vals (fromCBNCtx Γ) (fromCBNCtx Σ)
+  fromCBNTerms ∅ = ∅
+  fromCBNTerms (px • ts) = vThunk (fromCBN px) • fromCBNTerms ts
+
+  open import UNDEFINED
+  -- This is really supposed to be part of the plugin interface.
+  cbnToCompConst : ∀ {Σ τ} → Const Σ τ → CompConst (fromCBNCtx Σ) (cbnToCompType τ)
+  cbnToCompConst = reveal UNDEFINED
+
+  fromCBN (const c args) = cConst (cbnToCompConst c) (fromCBNTerms args)
+  fromCBN (var x) = cForce (vVar (fromVar cbnToValType x))
+  fromCBN (app s t) = cApp (fromCBN s) (vThunk (fromCBN t))
+  fromCBN (abs t) = cAbs (fromCBN t)
+
+  -- To satisfy termination checking, we'd need to inline fromCBV and weaken: fromCBV needs to produce a term in a bigger context.
+  -- But let's ignore that.
```

Comment:
Instead of inlining, we should see whether sized types from Agda 2.4.x helps. But that's for later, since we'd first have to switch to 2.4.